src/messaging/ExchangeContext.cpp

```diff
@@ -0,0 +1,390 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the ExchangeContext class.
+ *
+ */
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/CHIPProtocols.h>
+#include <protocols/common/CommonProtocol.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemTimer.h>
+
+using namespace chip::Encoding;
+using namespace chip::Inet;
+using namespace chip::System;
+
+namespace chip {
+
+// Special node id values.
+static constexpr NodeId kNodeIdNotSpecified = 0ULL;
+static constexpr NodeId kAnyNodeId          = 0xFFFFFFFFFFFFFFFFULL;
+
+static void DefaultOnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType,
+                                     PacketBuffer * payload)
+{
+    ChipLogError(ExchangeManager, "Dropping unexpected message %08" PRIX32 ":%d %04" PRIX16 " MsgId:%08" PRIX32, protocolId,
+                 msgType, ec->GetExchangeId(), packetHeader.GetMessageId());
+
+    PacketBuffer::Free(payload);
+}
+
+bool ExchangeContext::IsInitiator() const
+{
+    return mFlags.Has(ExFlagValues::kFlagInitiator);
+}
+
+bool ExchangeContext::IsResponseExpected() const
+{
+    return mFlags.Has(ExFlagValues::kFlagResponseExpected);
+}
+
+void ExchangeContext::SetInitiator(bool inInitiator)
+{
+    mFlags.Set(ExFlagValues::kFlagInitiator, inInitiator);
+}
+
+void ExchangeContext::SetResponseExpected(bool inResponseExpected)
+{
+    mFlags.Set(ExFlagValues::kFlagResponseExpected, inResponseExpected);
+}
+
+CHIP_ERROR ExchangeContext::SendMessage(uint16_t protocolId, uint8_t msgType, PacketBuffer * msgBuf, uint16_t sendFlags,
+                                        void * msgCtxt)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    PayloadHeader payloadHeader;
+
+    // Don't let method get called on a freed object.
+    VerifyOrDie(mExchangeMgr != nullptr && mRefCount != 0);
```

Comment:
Do we want to `abort()` (i.e. is it a fatal condition), or can we just return error, and continue?

Reply:
We can not recover from those error cases, we should abort.